src/events/attributes.rs

```diff
@@ -333,187 +278,2076 @@ impl<'a> From<(&'a str, &'a str)> for Attribute<'a> {
-        }
-
-        // value has quote if not html
-        match bytes.by_ref().find(|&(_, &b)| !is_whitespace(b)) {
-            Some((i, quote @ &b'\'')) | Some((i, quote @ &b'"')) => {
-                match bytes.by_ref().find(|&(_, &b)| b == *quote) {
-                    Some((j, _)) => {
-                        self.position = j + 1;
-                        return attr!(start_key..end_key, i + 1..j);
-                    }
-                    None => err!(Error::UnquotedValue(i)),
-                }
+            Ok(key) => key,
+        };
+
+        ////////////////////////////////////////////////////////////////////////
+
+        // Gets the position of quote and quote type
+        let (start_value, quote) = match iter.find(|(_, &b)| !is_whitespace(b)) {
+            // Input: `    key  =  "`
+            //                     ^
+            Some((s, b'"')) => (s + 1, b'"'),
+            // Input: `    key  =  '`
+            //                     ^
+            Some((s, b'\'')) => (s + 1, b'\''),
+
+            // Input: `    key  =  x`
+            //                     ^
+            // If HTML-like attributes is allowed, this is the start of the value
+            Some((s, _)) if self.html => {
+                // We do not check validity of attribute value characters as required
+                // according to https://html.spec.whatwg.org/#unquoted. It can be done
+                // during validation phase
+                let end = match iter.find(|(_, &b)| is_whitespace(b)) {
+                    // Input: `    key  =  value `
+                    //                     |    ^
+                    //                     s    e
+                    Some((e, _)) => e,
+                    // Input: `    key  =  value`
+                    //                     |    ^
+                    //                     s    e = len()
+                    None => slice.len(),
+                };
+                self.state = State::Next(end);
+                return Some(Ok(Attr::Unquoted(key, s..end)));
             }
-            Some((i, _)) if self.html => {
-                let j = bytes
-                    .by_ref()
-                    .find(|&(_, &b)| is_whitespace(b))
-                    .map_or(len, |(j, _)| j);
-                self.position = j;
-                return attr!(start_key..end_key, i..j);
+            // Input: `    key  =  x`
+            //                     ^
+            Some((s, _)) => {
+                self.state = State::SkipValue(s);
+                return Some(Err(AttrError::UnquotedValue(s)));
+            }
+
+            // Input: `    key  =  `
+            //                     ^
+            None => {
+                // Because we reach end-of-input, stop iteration on next call
+                self.state = State::Done;
+                return Some(Err(AttrError::ExpectedValue(slice.len())));
+            }
+        };
+
+        match iter.find(|(_, &b)| b == quote) {
```

Comment:
I also tried to replace this by
```rust
match memchr::memchr(quote, &slice[start_value..]) {
  ...
}
```
but that does not give a significance difference